text/0097-log-data-model.md

```diff
@@ -0,0 +1,971 @@
+existing log formats map to the fields defined below. If there are questions
+about the meaning of the field reviewing the examples may be helpful.
+
+Here is the list of fields in a log record:
+
+Field Name     |
+---------------|
+Timestamp      |
+TraceId        |
+SpanId         |
+TraceFlags     |
+SeverityText   |
+SeverityNumber |
+ShortName      |
+Body           |
+Resource       |
+Attributes     |
+
+Below is the detailed description of each field.
+
+### Field: `Timestamp`
+
+Type: Timestamp, uint64 nanoseconds since Unix epoch.
+
+Description: Time when the event occurred measured by the origin clock. This
+field is optional, it may be missing the timestamp is unknown.
+
+### Trace Context Fields
+
+#### Field: `TraceId`
+
+Type: byte sequence.
+
+Description: Optional request trace id. Can be set for logs that are part of
+request processing and have an assigned trace id.
+
+#### Field: `SpanId`
+
+Type: byte sequence.
+
+Description: Optional span id. Can be set for logs that are part of a particular
+processing span. If SpanId is present TraceId SHOULD be also present.
+
+#### Field: `TraceFlags`
+
+Type: byte.
+
+Description: Optional trace flag as defined in
+[W3C trace context](https://www.w3.org/TR/trace-context/#trace-flags)
+specification. At the time of writing the specification defines one flag - the
+SAMPLED flag.
+
+### Severity Fields
+
+#### Field: `SeverityText`
+
+Type: string.
+
+Description: the severity text (also known as log level). This is an optional
+field and is the original string representation as it is known at the source. If
+this field is missing and `SeverityNumber` is present then the short name that
+corresponds to the `SeverityNumber` can be used as a substitution.
+
+#### Field: `SeverityNumber`
+
+Type: number.
+
+Description: numerical value of the severity, normalized to values described in
+this document. This is an optional field. If `SeverityNumber` is missing and
+SeverityText is present then it may be assumed that `SeverityNumber` is equal
+to INFO (numeric 9) (see the meaning below).
+
+`SeverityNumber` is an integer number. Smaller numerical values correspond to
+less severe events (such as debug events), larger numerical values correspond to
+more severe events (such as errors and critical events). The following table
+defines the meaning of `SeverityNumber` value:
+
+SeverityNumber range|Range name|Meaning
+--------------------|----------|-------
+1-4                 |TRACE     |A fine-grained debugging event. Typically disabled in default configurations.
+5-8                 |DEBUG     |A debugging event.
+9-12                |INFO      |An informational event. Indicates that an event happened.
+13-16               |WARN      |A warning event. Not an error but is likely more important than an informational event.
+17-20               |ERROR     |An error event. Something went wrong.
+21-24               |FATAL     |A fatal error such as application or system crash.
```

Comment:
Might be difficult to work with Python since the logging level is defined as:

Level | Numeric value
-- | --
CRITICAL | 50
ERROR | 40
WARNING | 30
INFO | 20
DEBUG | 10
NOTSET | 0

Reply:
I was wondering what the numeric value was needed for in general as well. 

Or why so many?

In Erlang the severity levels are:

- emergency
- alert 
- critical 
- error 
- warning 
- notice 
- info 
- debug

which I believe correspond to syslog severities.

Reply:
I think Python will need to map to the smaller range defined by this data model and the custom level can be recorded in severity_text. I tried `logging.log(logging.WARN+1, 'some message')` and it prints `Level 31:root:some message`, so it has its internal notion of severity text already (`Level <N>`), which we can use.

Reply:
I think that using syslog levels will be better solution there.

Reply:
@hauleth I gave my opinion regarding applicability of Syslog as general log data model for OpenTelemetry elsewhere (my opinion is that it is not a good fit). To avoid repetition I will refrain from further comments on the same topic. Based on your comments it appears your opinion is different. I would be glad to review an alternate proposal that suggests to use Syslog as the data model if anyone wants to submit an alternate OTEP for that.

Reply:
@tigrannajaryan I'm more confused after your python example :(. If WARN is 13-16 and 30 in python why is the example `Level 31`?

I agree with you on not using the syslog data model but think the severity levels are a separate matter.

Reply:
Python logging library outputs `Level 31` as severity text when you use _Python's_ WARN+1 log level. Python's WARN is numerically 30. WARN+1 is thus numerically 31. We don't control Python logging library behavior and will use whatever the logging library believes to be the appropriate text for each particular logging level. In this case Python logging library believes that text that describes WARN+1 numeric level is `Level 31`. That's what we will have to use as `severity_text`, since by definition it is the original text of the severity (of log level).

Reply:
In practice I don't know how often people use custom log levels with Python logging library. Anecdotally, I worked for a few years on a couple Python codebases and never seen custom levels used, so this may be a minor edge case.

Reply:
Sorry, wasn't clear in what I was confused about. The code made sense, but I thought it was meant to show some form of integrating between the two severity numeric values (python and otels). But now I get that you were using it as an example to show the "severity text" could encompass python's log level text and integer value together.

I don't know how useful this would be, after it has left the node I would think users don't care if the level was 31 and would find more useful that all of their nodes (no matter the language) were outputting the same `severity_text` values.

I didn't think custom levels were the issue here. Whether it is 30 or 31 it is still not the same as the OpenTelemetry numeric value for a warning.

Reply:
@tsloughter if we go with 2 severity fields as proposed by this OTEP then the output of Python logging library will have to be converted to this data model by doing 2 things:
1. Mapping the numeric "log level" from Python logging library to `severity_number`. This is necessary because `severity_number` cannot represent the entire range of Python's log levels (and even if the range was large enough the semantics of numbers may not match anyway).
2. Using Python's "log level" texts as the value of `severity_text`. That matches this OTEPs definition of `severity_text`.

I agree that this does not look very nice to me. I am open to considering alternates, including and up to completely getting rid of severity fields from this data model until we have a proper solution.

Reply:
Can we at least keep `SeverityText` as a top-level field?

Reply:
I do not yet see any voices in favour of removing it so it stays for now :-)

Reply:
Great. Thanks!

Reply:
Resolving. Please reopen if necessary.